server/users.ts

```diff
@@ -407,48 +376,102 @@ class Connection {
 	onDisconnect() {
 		connections.delete(this.id);
 		if (this.user) this.user.onDisconnect(this);
-		this.user = /** @type {any} */ (null);
+		this.user = null!;
 	}
 
-	/**
-	 * @param {string} message
-	 */
-	popup(message) {
+	popup(message: string) {
 		this.send(`|popup|` + message.replace(/\n/g, '||'));
 	}
 
-	/**
-	 * @param {GlobalRoom | GameRoom | ChatRoom} room
-	 */
-	joinRoom(room) {
+	joinRoom(room: Room) {
 		if (this.inRooms.has(room.id)) return;
 		this.inRooms.add(room.id);
 		Sockets.roomAdd(this.worker, room.id, this.socketid);
 	}
-	/**
-	 * @param {GlobalRoom | GameRoom | ChatRoom} room
-	 */
-	leaveRoom(room) {
+	leaveRoom(room: Room) {
 		if (this.inRooms.has(room.id)) {
 			this.inRooms.delete(room.id);
 			Sockets.roomRemove(this.worker, room.id, this.socketid);
 		}
 	}
 	toString() {
-		return (this.user ? this.user.userid + '[' + this.user.connections.indexOf(this) + ']' : '[disconnected]') + ':' + this.ip + (this.protocol !== 'websocket' ? ':' + this.protocol : '');
+		let buf = this.user ? `${this.user.userid}[${this.user.connections.indexOf(this)}]` : `[disconnected]`;
+		buf += `:${this.ip}`;
+		if (this.protocol !== 'websocket') buf += `:${this.protocol}`;
+		return buf;
 	}
 }
 
-/** @typedef {[string, string, Connection]} ChatQueueEntry */
-
-const SETTINGS = ['isSysop', 'isStaff', 'blockChallenges', 'blockPMs', 'ignoreTickets', 'lastConnected', 'inviteOnlyNextBattle'];
+type ChatQueueEntry = [string, string, Connection];
+const SETTINGS = [
+	'isSysop', 'isStaff', 'blockChallenges', 'blockPMs',
+	'ignoreTickets', 'lastConnected', 'inviteOnlyNextBattle',
+];
 
 // User
 class User extends Chat.MessageContext {
-	/**
-	 * @param {Connection} connection
-	 */
-	constructor(connection) {
+	user: User;
+	mmrCache: {[format: string]: number};
+	guestNum: number;
+	name: string;
+	named: boolean;
+	registered: boolean;
+	userid: ID;
+	group: string;
+	avatar: string | number;
+
+	connected: boolean;
+	connections: Connection[];
+	latestHost: string;
+	latestHostType: string;
+	ips: {[k: string]: number};
+	latestIp: string;
+	locked: string | ID | null;
+	semilocked: string | null;
+	namelocked: string | ID | null;
+	permalocked: string | ID | null;
+	// actually {[k: ID]: string}, but typescript doesn't like that
```

Comment:
i like doing `{[id: string]: string}` to add slightly more documentation that the key is intended to be an ID, or even `{[id: /* ID */ string]: string}`.